File: Day_4/Sliding_Window/code/Solution_2/sol.py
def length_of_longest_substring_k_distinct(s: str, k: int) -> int:
    if k == 0 or not s:
        return 0

    left = 0
    max_len = 0
    char_count = {}

    for right, char in enumerate(s):
        char_count[char] = char_count.get(char, 0) + 1

        while len(char_count) > k:
            left_char = s[left]
            char_count[left_char] -= 1
            if char_count[left_char] == 0:
                del char_count[left_char]
            left += 1

        max_len = max(max_len, right - left + 1)

    return max_len

# Example usage:
s = "eceba"
k = 2
print(length_of_longest_substring_k_distinct(s, k))  # Output: 3

File: Day_4/Sliding_Window/code/Solution_4/sol.cpp
#include <vector>
#include <algorithm>
using namespace std;

int longestOnes(vector<int>& nums, int k) {
    int left = 0, right = 0, maxLen = 0, zeros = 0;

    while (right < nums.size()) {
        if (nums[right] == 0) zeros++;

        while (zeros > k) {
            if (nums[left] == 0) zeros--;
            left++;
        }

        maxLen = max(maxLen, right - left + 1);
        right++;
    }

    return maxLen;
}
